libs/utils/energy_model.py

```diff
@@ -0,0 +1,572 @@
+                      default.
+        :param idle_states: List of CPU frequencies. Got from
+                            :meth:`guess_idle_states` by default.
+
+        :returns: Dict with power in bogo-Watts (bW), with contributions from
+                  each system component keyed with a tuple of the CPUs
+                  comprising that component (i.e. :attr:EnergyModelNode.cpus)
+
+                  ::
+
+                    {
+                        (0,)    : 10,
+                        (1,)    : 10,
+                        (1, 2)  : 5,
+                    }
+
+                  This represents CPUs 0 and 1 each using 10bW and their shared
+                  resources using 5bW for a total of 25bW.
+        """
+        if len(util_distrib) != len(self.cpus):
+            raise ValueError(
+                'util_distrib length ({}) must equal CPU count ({})'.format(
+                    len(util_distrib), len(self.cpus)))
+
+        if freqs is None:
+            freqs = self.guess_freqs(util_distrib)
+        if idle_states is None:
+            idle_states = self.guess_idle_states(util_distrib)
+
+        cpu_active_time = []
+        for cpu, node in enumerate(self.cpu_nodes):
+            assert (cpu,) == node.cpus
+            cap = node.active_states[freqs[cpu]].capacity
+            cpu_active_time.append(min(float(util_distrib[cpu]) / cap, 1.0))
+
+        return self._estimate_from_active_time(cpu_active_time,
+                                               freqs, idle_states, combine=True)
+
+    def get_optimal_placements(self, capacities):
+        """Find the optimal distribution of work for a set of tasks
+
+        Find a list of candidates which are estimated to be optimal in terms of
+        power consumption, but that do not result in any CPU becoming
+        over-utilized.
+
+        If no such candidates exist, i.e. the system being modeled cannot
+        satisfy the workload's throughput requirements, an
+        :class:`EnergyModelCapacityError` is raised. For example, if e was an
+        EnergyModel modeling two CPUs with capacity 1024, this error would be
+        raised by:
+
+        ::
+
+          e.get_optimal_placements({"t1": 800, "t2": 800, "t3: "800"})
+
+        This estimation assumes an ideal system of selecting OPPs and idle
+        states for CPUs.
+
+        .. note::
+            This is a brute force search taking time exponential wrt. the number
+            of tasks.
+
+        :param capacities: Dict mapping tasks to expected utilization
+                           values. These tasks are assumed not to change; they
+                           have a single static utilization value. A set of
+                           single-phase periodic RT-App tasks is an example of a
+                           suitable workload for this model.
+        :returns: List of ``util_distrib`` s representing distribution of work
+                  under optimal task placments see
```

Comment:
mm.. now I'm thinking about....

:balloon:  [**CI FOR SPELLING**](https://sphinxcontrib-spelling.readthedocs.io/en/latest/) :balloon: